### PR TITLE
mainloop-call: fix hang caused by the previous fix for pending mainloop-calls

### PR DESCRIPTION
**Pull request by chucksilvers**
Now that we've had the previous fix from #4258 in production for a while, we've started to see that syslog-ng is now occasionally hanging, and the hangs are unfortunately caused by that previous fix.  

Here are the stack traces of the two threads that cause the hang:

```
(gdb) i thr
  Id   Target Id                  Frame 
* 1    LWP 101322 of process 1003 _umtx_op_err ()
    at /usr/src/lib/libthr/arch/amd64/amd64/_umtx_op_err.S:40
  2    LWP 101323 of process 1003 _kevent () at _kevent.S:4
  3    LWP 976246 of process 1003 _umtx_op_err ()
    at /usr/src/lib/libthr/arch/amd64/amd64/_umtx_op_err.S:40


(gdb) bt
#0  _umtx_op_err ()
    at /usr/src/lib/libthr/arch/amd64/amd64/_umtx_op_err.S:40
#1  0x000015e1cb8f324c in __thr_umutex_lock (mtx=0x15e1d480a688, id=0x18bca)
    at /usr/src/lib/libthr/thread/thr_umtx.c:82
#2  0x000015e1cb8eba7b in mutex_lock_sleep (curthread=0x15e1d21a6000, m=0x15e1d480a688, abstime=<optimized out>)
    at /usr/src/lib/libthr/thread/thr_mutex.c:702
#3  mutex_lock_common (m=0x15e1d480a688, abstime=abstime@entry=0x0, cvattach=0x0, rb_onlist=0x0)
    at /usr/src/lib/libthr/thread/thr_mutex.c:736
#4  0x000015e1cb8eae16 in __Tthr_mutex_lock (mutex=mutex@entry=0x15e1d21df000)
    at /usr/src/lib/libthr/thread/thr_mutex.c:755
#5  0x000015e1c5dcbc22 in ___mutex_lock (mutex=0x15e1d21df000) at ./mutex.h:53
#6  iv_work_submit_pool (this=0x15e1c5de3548 <main_loop_io_workers>, work=0x15e1d229db70) at iv_work.c:315
#7  iv_work_pool_submit_work (this=0x15e1c5de3548 <main_loop_io_workers>, work=0x15e1d229db70) at iv_work.c:394
#8  0x000015e1c5dcaf2a in iv_run_tasks (st=st@entry=0x15e1d21e4000) at iv_task.c:48
#9  0x000015e1c5dcda77 in iv_main () at iv_main_posix.c:99
#10 0x000015e1c5d756f4 in main_loop_run (self=0x15e1c5de2ae8 <main_loop>) at lib/mainloop.c:677
#11 0x000015d9a4a657bd in main (argc=<optimized out>, argv=0x15e1c4e510b8) at syslog-ng/main.c:335


(gdb) bt
#0  _umtx_op_err ()
    at /usr/src/lib/libthr/arch/amd64/amd64/_umtx_op_err.S:40
#1  0x000015e1cb8f3621 in _thr_umtx_timedwait_uint (mtx=0x15e1db0220e0, id=0x0, clockid=<optimized out>, 
    abstime=<optimized out>, shared=<optimized out>)
    at /usr/src/lib/libthr/thread/thr_umtx.c:236
#2  0x000015e1cb8e5b7f in cond_wait_user (cvp=0x15e1f5737060, mp=0x15e1d481cc88, abstime=0x0, cancel=0x1)
    at /usr/src/lib/libthr/thread/thr_cond.c:320
#3  cond_wait_common (cond=cond@entry=0x15e1e0e03008, mutex=<optimized out>, abstime=abstime@entry=0x0, 
    cancel=cancel@entry=0x1)
    at /usr/src/lib/libthr/thread/thr_cond.c:380
#4  0x000015e1cb8e5e01 in __thr_cond_wait (cond=0x15e1db0220e0, cond@entry=0x15e1e0e03008, mutex=0xf)
    at /usr/src/lib/libthr/thread/thr_cond.c:395
#5  0x000015e1caafb855 in g_cond_wait (cond=<optimized out>, mutex=mutex@entry=0x15e1c5de3470 <main_task_lock>)
    at ../glib/gthread-posix.c:794
#6  0x000015e1c5d75c85 in main_loop_wait_for_pending_call_to_finish () at lib/mainloop-call.c:73
#7  0x000015e1c5d75d0b in main_loop_call_thread_deinit () at lib/mainloop-call.c:145
#8  0x000015e1c5d5aa67 in app_thread_stop () at lib/apphook.c:347
#9  0x000015e1c5d760ab in main_loop_worker_thread_stop () at lib/mainloop-worker.c:205
#10 0x000015e1c5dcc1b1 in __iv_work_thread_die (thr=0x15e1e59991c0) at iv_work.c:75
#11 0x000015e1c5dcc14e in iv_work_thread_idle_timeout (_thr=0x15e1e59991c0) at iv_work.c:154
#12 0x000015e1c5dcb309 in iv_run_timers (st=st@entry=0x15e1e0e0b000) at iv_timer.c:124
#13 0x000015e1c5dcda6f in iv_main () at iv_main_posix.c:98
#14 0x000015e1c5dcbea7 in iv_work_thread (_thr=0x15e1e59991c0) at iv_work.c:185
#15 0x000015e1c5dcf04a in iv_thread_handler (_thr=0x15e1d2385a20) at iv_thread_posix.c:112
#16 0x000015e1cb8e6cf8 in thread_start (curthread=0x15e1d232d400)
    at /usr/src/lib/libthr/thread/thr_create.c:292
```

Here is the mutex that the main-loop thread is blocked on:

```
(gdb) frame 5
#5  0x000015e1c5dcbc22 in ___mutex_lock (mutex=0x15e1d21df000) at ./mutex.h:53
53	./mutex.h: No such file or directory.
(gdb) p mutex
$4 = (___mutex_t *) 0x15e1d21df000
(gdb) p *mutex
$5 = (___mutex_t) 0x15e1d480a688
(gdb) p **mutex
$6 = {
  m_lock = {
    m_owner = 0x800ee576,
    m_flags = 0x0,
    m_ceilings = {0x0, 0x0},
    m_rb_lnk = 0x0,
    m_spare = {0x0, 0x0}
  },
  m_flags = 0x1,
  m_count = 0x0,
  m_spinloops = 0x0,
  m_yieldloops = 0x0,
  m_ps = 0x0,
  m_qe = {
    tqe_next = 0x0,
    tqe_prev = 0x15e1d232d5a8
  },
  m_pqe = {
    tqe_next = 0x0,
    tqe_prev = 0x15e1d232d5b8
  },
  m_rb_prev = 0x0
}
(gdb) p/d 0xee576
$7 = 976246
```

So here we see that the main-loop thread is blocked waiting for a mutex owned by the worker thread, and the worker thread is blocked waiting for the main-loop thread to be done processing the mainloop-call.  The mutex in question is the iv_work_pool's lock:

```
(gdb) p this
$10 = (struct iv_work_pool *) 0x15e1c5de3548 <main_loop_io_workers>
(gdb) p pool
$11 = (struct work_pool_priv *) 0x15e1d21df000
(gdb) p pool->lock
$12 = (___mutex_t) 0x15e1d480a688
```

and this mutex is taken by the dying worker thread a couple frames up in the stack in iv_work_thread_idle_timeout:

```
static void iv_work_thread_idle_timeout(void *_thr)
{
        struct work_pool_thread *thr = _thr;
        struct work_pool_priv *pool = thr->pool;

        ___mutex_lock(&pool->lock);
        
        if (thr->kicked) {
                thr->idle_timer.expires = iv_now;
                thr->idle_timer.expires.tv_sec += 10;
                iv_timer_register(&thr->idle_timer);
        } else {
                iv_list_del_init(&thr->list);
                __iv_work_thread_die(thr);
        }

        ___mutex_unlock(&pool->lock);
}
```

This pull request fixes this hang by moving the mainloop-call TLS state from the worker's TLS structure to a separate structure that the worker's TLS points to. Then rather than waiting for a pending mainloop-call to complete when a worker thread wants to exit, just hand off responsibility for freeing the separate mainloop-call structure when the mainloop-call is complete from the exiting worker thread to the mainloop thread itself.

### CODE REVIEW

**Comment by kira-syslogng** [comment]
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
"add to whitelist" add author of a Pull Request to whitelist (globally, be careful, it means this user can trigger kira for any PR)
do nothing -> CI won't start)

**Comment by kira-syslogng** [comment]
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
"add to whitelist" add author of a Pull Request to whitelist (globally, be careful, it means this user can trigger kira for any PR)
do nothing -> CI won't start)

**Comment by alltilla** [comment]
@kira-syslogng test this please

**Comment by kira-syslogng** [comment]
Build FAILURE

File: lib/mainloop-call.c [hunk 1]
```diff
@@ -135,17 +150,26 @@ main_loop_call_handler(gpointer user_data)
 void
 main_loop_call_thread_init(void)
 {
-  g_cond_init(&call_info.cond);
-  g_mutex_init(&call_info.lock);
+  call_info = g_malloc(sizeof (MainLoopTaskCallSite));
```
Comment (alltilla) [inline lib/mainloop-call.c:153]: Let's 0 init all the fields:

```suggestion
  call_info = g_new0(MainLoopTaskCallSite, 1);
```

**Review by alltilla — CHANGES_REQUESTED** [review]
Thank you very much for the thorough analysis and proposed fix!

I agree with your points, there is a deadlock in the following case:
1. A worker thread calls `main_loop_call()` with `wait == FALSE`, so it instantly returns.
2. There is a task (let's call it `task_0`) already in the main event loop, so the task (`task_wt`) gets registered after that one.
3. After 10 seconds `task_0` has not been processed yet, so the worker thread gets deinited, but as `task_wt` is still pending, the deinit waits for it in `cond_wait()`. The ivykis worker thread lock is held.
4. `task_0` gets to be processed, and it wants to create a job in the worker threads, but it cannot, as it needs to grab the ivykis worker thread lock, but it is held by the deiniting worker thread.

Waiting in deinit is problematic, so I agree with you, in some cases the main loop should free the `MainLoopTaskCallSite`.

I think there is a small problem in your proposed solution, where not all fields of `MainLoopTaskCallSite` gets initialized to 0, while it was done so implicitly when it was in TLS before, this causes the CI failure. Otherwise LGTM!

**Comment by alltilla** [comment]
@kira-syslogng test this please

**Comment by alltilla** [comment]
@kira-syslogng do stresstest

**Comment by alltilla** [comment]
@kira-syslogng test this please

**Comment by alltilla** [comment]
@kira-syslogng do stresstest

**Comment by chucksilvers** [comment]
It took me a couple tries but I think I got github to update my PR branch with what I wanted.

I saw that the CI had failed, but I didn't see any indication on the https://github.com/syslog-ng/syslog-ng/actions/runs/5395637688/jobs/9804582093?pr=4524 page of why it had failed, all it says is "Error:".  Where did you find the information that the CI failure was due to parts of the new structure being left uninitialized?

**Comment by alltilla** [comment]
Thank you!

I think only authorized users can see the output of the CI runs of PRs. However in your fork you could probably access them.

I saw hangs in the E2E tests, and after a bit of investigation I found the culprit. Valgrind would have probably shown uninitialized reads, too, but have not checked it. After I have fixed it locally the tests passed for me.

**Comment by alltilla** [comment]
@kira-syslogng do stresstest

**Comment by alltilla** [comment]
@kira-syslogng test this please

**Comment by chucksilvers** [comment]
I fixed the style thing now too.  For some reason the style-checker didn't complain about my change when I ran it before (and it complained about lots of other files), but running it again now it complained only about my change.

**Comment by alltilla** [comment]
Thank you, perfect! :)

I will start all the CI jobs, and merge the PR if they succeed. Our stress test does not start for some reason, if I cannot make it work, I will have to check it tomorrow.

**Comment by alltilla** [comment]
@kira-syslogng do stresstest

**Comment by alltilla** [comment]
@kira-syslogng do stresstest

**Comment by alltilla** [comment]
Manually started the stress test. Run ID: 1960

**Comment by alltilla** [comment]
@kira-syslogng do stresstest

**Comment by alltilla** [comment]
Please ignore the failed macOS CI run. I have pushed a `fix-mainloop-call-hang` branch to `syslog-ng/syslog-ng` and for some reason its CI runs appeared here. The macOS run fails because of a problem which is fixed on master, and is not relevant to this change.

**Review by alltilla — APPROVED** [review]

**Comment by kira-syslogng** [comment]
Kira-stress-test: Build SUCCESS

**Comment by alltilla** [comment]
@chucksilvers Thank you again for your contribution! Sorry for the bumpy road, usually it is not this difficult to merge PRs, we got unlucky this time.

**Pull request merged**